### PR TITLE
Optimize `slice_iter.copied().next_chunk()`

### PR DESCRIPTION
**Pull request by the8472**
```
OLD: 
test iter::bench_copied_array_chunks                               ... bench:         371 ns/iter (+/- 7)
NEW:
test iter::bench_copied_array_chunks                               ... bench:          31 ns/iter (+/- 0)
```

The default `next_chunk` implementation suffers from having to assemble the array byte by byte via `next()`, checking the `Option<&T>` and then dereferencing `&T`. The specialization copies the chunk directly from the slice.

### CODE REVIEW

**Comment by vacuus** [comment]
Just my two cents, use `raw_array`/`raw_arr` instead of `raw_ary` for the variable name. I don't think that'll go over a formatting line length threshold, anyway. Also, the safety comment should have a `;` instead of `,`.

File: library/core/src/iter/adapters/copied.rs [hunk 1]
```diff
@@ -166,3 +178,65 @@ where
...
+            }
+        }
+
+        // SAFETY: `len` is larger than the array size. Copy a fixed amount here to fully initialize
+        // the array.
+        return unsafe {
+            ptr::copy_nonoverlapping(self.as_ref().as_ptr(), raw_ary.as_mut_ptr() as *mut T, N);
+            let _ = self.advance_by(N);
+            Ok(MaybeUninit::array_assume_init(raw_ary))
+        };
```
Comment (m-ou-se) [inline library/core/src/iter/adapters/copied.rs:240]:
```suggestion
        unsafe {
            ptr::copy_nonoverlapping(self.as_ref().as_ptr(), raw_ary.as_mut_ptr() as *mut T, N);
            let _ = self.advance_by(N);
            Ok(MaybeUninit::array_assume_init(raw_ary))
        }
```

File: library/core/src/iter/adapters/copied.rs [hunk 2]
```diff
@@ -166,3 +178,65 @@ where
...
+        array::iter_next_chunk(&mut self.map(|e| *e))
+    }
+}
+
+impl<'a, const N: usize, T: 'a> SpecNextChunk<'a, N, T> for crate::slice::Iter<'a, T>
+where
+    T: Copy,
+{
+    fn spec_next_chunk(&mut self) -> Result<[T; N], array::IntoIter<T, N>> {
+        let mut raw_ary = MaybeUninit::uninit_array();
```
Comment (m-ou-se) [inline library/core/src/iter/adapters/copied.rs:204]:
```suggestion
        let mut raw_array = MaybeUninit::uninit_array();
```

**Review by m-ou-se — APPROVED** [review]
r=me :)

**Pull request merged**